sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ToStringBase.scala

```diff
@@ -0,0 +1,414 @@
+  private lazy val timestampNTZFormatter = TimestampFormatter.getFractionFormatter(ZoneOffset.UTC)
+
+  // The brackets that are used in casting structs and maps to strings
+  protected def leftBracket: String
+  protected def rightBracket: String
+
+  // The string value to use to represent null elements in array/struct/map.
+  protected def nullString: String
+
+  protected def useDecimalPlainString: Boolean
+
+  protected def useHexFormatForBinary: Boolean
+
+  // Makes the function accept Any type input by doing `asInstanceOf[T]`.
+  @inline private def acceptAny[T](func: T => Any): Any => Any = i => func(i.asInstanceOf[T])
+
+  // Returns a function to convert a value to pretty string. The function assumes input is not null.
+  protected final def castToString(from: DataType): Any => Any = from match {
+    case CalendarIntervalType =>
+      acceptAny[CalendarInterval](i => UTF8String.fromString(i.toString))
+    case BinaryType if useHexFormatForBinary =>
+      acceptAny[Array[Byte]](binary => UTF8String.fromString(StringUtils.getHexString(binary)))
+    case BinaryType =>
+      acceptAny[Array[Byte]](UTF8String.fromBytes)
+    case DateType =>
+      acceptAny[Int](d => UTF8String.fromString(dateFormatter.format(d)))
+    case TimestampType =>
+      acceptAny[Long](t => UTF8String.fromString(timestampFormatter.format(t)))
+    case TimestampNTZType =>
+      acceptAny[Long](t => UTF8String.fromString(timestampNTZFormatter.format(t)))
+    case ArrayType(et, _) =>
+      acceptAny[ArrayData](array => {
+        val builder = new UTF8StringBuilder
+        builder.append("[")
+        if (array.numElements > 0) {
+          val toUTF8String = castToString(et)
+          if (array.isNullAt(0)) {
+            if (nullString.nonEmpty) builder.append(nullString)
+          } else {
+            builder.append(toUTF8String(array.get(0, et)).asInstanceOf[UTF8String])
+          }
+          var i = 1
+          while (i < array.numElements) {
+            builder.append(",")
+            if (array.isNullAt(i)) {
+              if (nullString.nonEmpty) builder.append(" " + nullString)
+            } else {
+              builder.append(" ")
+              builder.append(toUTF8String(array.get(i, et)).asInstanceOf[UTF8String])
+            }
+            i += 1
+          }
+        }
+        builder.append("]")
+        builder.build()
+      })
+    case MapType(kt, vt, _) =>
+      acceptAny[MapData](map => {
+        val builder = new UTF8StringBuilder
+        builder.append(leftBracket)
+        if (map.numElements > 0) {
+          val keyArray = map.keyArray()
+          val valueArray = map.valueArray()
+          val keyToUTF8String = castToString(kt)
+          val valueToUTF8String = castToString(vt)
+          builder.append(keyToUTF8String(keyArray.get(0, kt)).asInstanceOf[UTF8String])
+          builder.append(" ->")
+          if (valueArray.isNullAt(0)) {
+            if (nullString.nonEmpty) builder.append(nullString)
```

Comment:
Sorry for late review.

This seems to cause a bug because previously we had the following.
```
if (!legacyCastToStr) builder.append(" NULL")
```

In other words, we have a space.
```
scala> spark.version
res9: String = 3.4.0

scala> sql("select map('k', null)").show()
+------------+
|map(k, NULL)|
+------------+
| {k -> null}|
+------------+
```

Now, master branch shows the following.
```
scala> spark.version
res9: String = 3.5.0-SNAPSHOT

scala> sql("select map('k', null)").show()
+------------+
|map(k, NULL)|
+------------+
|  {k ->NULL}|
+------------+
```

Reply:
I made a follow-up for this.
- https://github.com/apache/spark/pull/41432